src/db/design/vote_registration_query.md

```diff
@@ -0,0 +1,268 @@
+SELECT DISTINCT ON (metadata->'2') * FROM (
+    SELECT * FROM (
+        SELECT  tx.hash, tx_id, metadata, signature,
+                block.slot_no AS block_slot_no,
+                (
+                    CASE WHEN (metadata->>'5')~E'^\\d+$' THEN
+                        CAST ((metadata->>'5') AS INTEGER)
+                    WHEN metadata->>'5' IS NULL THEN
+                        0
+                    ELSE
+                        -1
+                    END
+                ) AS purpose
+        FROM meta_table
+                 INNER JOIN tx ON tx.id = meta_table.tx_id
+                 INNER JOIN sig_table USING (tx_id)
+                 INNER JOIN block ON block.id = tx.block_id
+        WHERE metadata ? '1' AND metadata ? '2' AND metadata ? '3' AND metadata ? '4' AND
+              signature ? '1' AND
+              block.slot_no > 0 AND block.slot_no <= 62510369
+        ORDER BY metadata -> '2', metadata -> '4' DESC, tx_id ASC
+    ) as all_registrations WHERE purpose = 0
+) as purpose_registrations;
+```
+
+This query:
+
+1. Can get registrations on any period, defined between two block numbers, inclusively:
+
+   ```sql
+   block.slot_no > 0 AND block.slot_no <= 62510369
+   ```
+
+   must be parameterized in the query above.
+2. It also filters out ONLY the latest registration for each stake key, as per the original iterative logic. (Eliminates the need to do it in code iteratively).
+
+    ```sql
+    SELECT DISTINCT ON (metadata->'2') * FROM (
+        --- ...
+        ORDER BY metadata -> '2', metadata -> '4' DESC, tx_id ASC
+    ```
+
+    Is the part of the query which filters that, and does NOT need parameterization.
+
+3. Ensures that the metadata and signature have the minimum number of required
+   fields, which pre-filters some badly formed transactions from the results:
+
+    ```sql
+    WHERE metadata ? '1' AND metadata ? '2' AND metadata ? '3' AND metadata ? '4' AND
+          signature ? '1' AND
+    ```
+
+    Is the part of the query which does that, and does not need to be parameterized.
+
+4. Adds a new column called `purpose` which is the voting purpose of the
+   registration in alignment with CIP-36. The query is then filtered on the
+   `purpose`, so that ONLY registrations for the required purpose are found.
+   This properly does not exclude multiple registrations for multiple purposes.
+
+    ```sql
+    (
+        CASE WHEN (metadata->>'5')~E'^\\d+$' THEN
+            CAST ((metadata->>'5') AS INTEGER)
+        WHEN metadata->>'5' IS NULL THEN
+            0
+        ELSE
+            -1
+        END
+    ) AS purpose
+    ```
+
+    This SQL creates the purpose column. IF the `5` field is present in the
+    metadata, AND its a properly formed integer, it becomes the purpose (as an
+    integer). If it is NOT present, it is defaulted to purpose 0 (Cip-15->36
+    upgrade). Otherwise the purpose is set to -1 to indicate the registration
+    record is invalid and should be ignored.
+
```

Comment:
Document that `~` is a regex match operation and `E'` is an "Escape String Constant" ref: https://www.postgresql.org/docs/current/sql-syntax-lexical.html